front/lib/api/assistant/actions/mcp.ts

```diff
@@ -0,0 +1,365 @@
+import logger from "@app/logger/logger";
+
+interface MCPActionBlob {
+  id: ModelId;
+  agentMessageId: ModelId;
+  params: Record<string, unknown>;
+  tokensCount: number | null;
+  output: string | null;
+  functionCallId: string | null;
+  functionCallName: string | null;
+  step: number;
+}
+
+const CONTEXT_SIZE_DIVISOR_FOR_OUTPUT = 2;
+
+// This action uses the MCP protocol to communicate
+export class MCPAction extends BaseAction {
+  readonly agentMessageId: ModelId;
+  readonly params: Record<string, unknown>; // Hold the inputs for the action.
+  readonly tokensCount: number | null = null;
+  readonly output: string | null;
+  readonly functionCallId: string | null;
+  readonly functionCallName: string | null;
+  readonly step: number = -1;
+  readonly type: MCPActionType["type"] = "mcp_action";
+
+  constructor(blob: MCPActionBlob) {
+    // Why we need the type here since we already have it above ?
+    super(blob.id, "mcp_action");
+
+    this.agentMessageId = blob.agentMessageId;
+    this.params = blob.params;
+    this.tokensCount = blob.tokensCount;
+    this.output = blob.output;
+    this.functionCallId = blob.functionCallId;
+    this.functionCallName = blob.functionCallName;
+    this.step = blob.step;
+  }
+
+  renderForFunctionCall(): FunctionCallType {
+    if (!this.functionCallId) {
+      throw new Error("MCPAction: functionCallId is required");
+    }
+    if (!this.functionCallName) {
+      throw new Error("MCPAction: functionCallName is required");
+    }
+
+    return {
+      id: this.functionCallId,
+      name: this.functionCallName,
+      arguments: JSON.stringify(this.params),
+    };
+  }
+
+  async renderForMultiActionsModel({
+    model,
+  }: {
+    conversation: ConversationType;
+    model: ModelConfigurationType;
+  }): Promise<FunctionMessageTypeModel> {
+    if (!this.functionCallName) {
+      return {
+        role: "function" as const,
+        name: "missing_function_call_name",
+        function_call_id: this.functionCallId ?? "missing_function_call_id",
+        content: `Error: the content was not tokenized`,
+      };
+    }
+
+    if (!this.functionCallId) {
+      return {
+        role: "function" as const,
+        name: this.functionCallName,
+        function_call_id: "missing_function_call_id",
+        content: `Error: the content was not tokenized`,
+      };
+    }
+
+    if (this.tokensCount === null) {
+      return {
+        role: "function" as const,
+        name: this.functionCallName,
+        function_call_id: this.functionCallId,
+        content: `Error: the content was not tokenized`,
+      };
+    }
+
+    if (
+      this.tokensCount >
+      model.contextSize / CONTEXT_SIZE_DIVISOR_FOR_OUTPUT
+    ) {
+      return {
+        role: "function" as const,
+        name: this.functionCallName,
+        function_call_id: this.functionCallId,
+        content: `Error: Action output has too many tokens to be included.`,
```

Comment:
Should we rather truncate (and append somnething like "[X more characters truncated]") ?